### DISCUSSION TITLE
Powerwall Status Monitor and API Request Dumper

### DISCUSSION THREAD

**Discussion by mcbirse** [jasonacox/Powerwall-Dashboard#249] [Show and tell]
I wrote a shell script way back, before I even knew about Powerwall-Dashboard.

I thought it may become obsolete for me given Grafana's alert capabilities, but it has not, and I still find it very useful to be running in the background as a system service on my Alpine Linux host.

It uses curl to poll the gateway with some API requests to get status of grid, battery percentage, and firmware version, and then sends an alert via e-mail on changes or certain conditions (low battery, etc.).

I originally wrote it as I was frustrated with how long it took the Tesla app to notify you of grid outages. With this script running, I'll get an e-mail within 5-10 seconds. It will also notify you of firmware updates within about a minute after update.

I have refined and added features over time. For instance, it will now poll the gateway via the pyPowerwall proxy, but will detect if that is not running/responding, and fall back to direct requests to the gateway.

So, it can run independently of Powerwall-Dashboard, which can be useful. But, it will still utilise the pyPowerwall proxy when available, so as not to overload the gateway with requests from lots of different sources and also utilise the proxies in-built caching.

It's also handy as a API request dumper via simple shell commands.

I have submitted a PR to add this script to the tools, as I believe others may find it useful also. Refer here once merged: [Details and Instructions](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/tools/pwstatus/)

My setup and configuration to install the script as a system service is quite specific to Alpine Linux however, so if anyone has further instructions to install on other platforms please feel free to add to this discussion or submit a PR to update the README!

Raspberry Pi seems to be commonly used for Powerwall-Dashboard and I assume the set up would be different (I have not used it myself), so some guidance for this platform may be helpful to the wider audience perhaps. 

Feedback welcome - would love to hear if this is a useful addition, or if you find any problems or have questions please let us know.

**Comment by aristobrat** [comment]
Hi @mcbirse!  I'm working to get this setup and am running into this when testing in request mode:

`pi@homebridge:~/Powerwall-Dashboard/tools/pwstatus $ ./pwstatus.sh api/system_status/grid_status api/system_status/soe | jq -s .
./pwstatus.sh: 324: Bad substitution`

My kit is a Raspberry Pi 3 Model B Rev 1.2 running the Homebridge Raspberry PI image (which looks to be a Raspian-based Raspberry Pi image with Homebridge preinstalled).

My /etc/pwstatus.conf looks OK to me. I can manually logon to my PW using the IP/credentials supplied there. The PWPROXY URL appears to work as well (I see the animation OK).

curl was preinstalled (below) and I manually installed a mta/mailx and can send email OK.

Any help on what to check would be greatly appreciated!

`pi@homebridge:~/Powerwall-Dashboard/tools/pwstatus $ curl --version
curl 7.74.0 (arm-unknown-linux-gnueabihf) libcurl/7.74.0 OpenSSL/1.1.1w zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.43.0 librtmp/2.3
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets`

**Reply by mcbirse** [reply]
Hi @aristobrat - that is strange. I am guessing there must be some compatibility difference with the shell interpreter. I ran your example and it should output like below.

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/108725631/91bf87d3-653a-4a23-94ed-84611210420b)

Can you try the following and see if it resolves it?

Edit the script and go to line 324.

At the line below starting with "if" change the `=` into 2 consecutive `==` like below.

```sh
        for stat in $@
        do
            if [ "${stat:0:1}" == "/" ]
            then
                stat="${stat:1}"
            fi
```

Let us know if that resolves the issue.

If it does, I will update the script and change all condition checks like that to `==` which perhaps is more compatible.

Alternatively, it probably works fine with bash I imagine, so maybe I should just change the script to use bash by default...

**Reply by aristobrat** [reply]
I didn't have any luck with changing the `=` into 2 consecutive `==`, but changing the script to use bash worked (at least for `./pwstatus.sh --monitor`). I did need to `apt install bc`, but that was easy enough. 

Thanks for the quick assist!

**Reply by mcbirse** [reply]
Thanks for the update and glad it is working!

Yes, I was wondering if the issue was with the use of a single `=` or the parameter expansion / variable slicing - clearly the latter. The default shell interpreter on Raspberry Pi 3 must not support it? Odd.

I think I will update the script to just default to bash. Much less likely for people to have issues, and bash should be able to be easily installed on most systems if missing.
